Refactor Preview for WorkspaceEdit with `needsConfirmation` flag

BoykoAlex opened the pull request:
`WorkspaceEdit` has a map of change annotations where any change annotation is associated with edits from the workspace edit. If any of change annotations has `needsConfirmation` flag on then VSCode would bring up the refactor preview with annotated changes unchecked.

This PR proposes similar (but not the same, there are limitations) behaviour into Eclipse.
If there is at lease one change annotation that requires confirmation then refactor wizard is opened with the Preview page opened. All changes would be checked marked as I couldn't find how to control this piece. Another limitation is Eclipse text edits are in reverse order (they are applied in reverse order for a reason) but the preview doesn't allow to set the tree sorter on the preview tree.

@rubenporras @mickaelistria @martinlippert feedback, ideas are welcomed :-)

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java:
```diff
@@ -925,25 +933,79 @@ public static void applyWorkspaceEdit(WorkspaceEdit wsEdit, String label) {
-				});
+
+		if (wsEdit.getChangeAnnotations() != null && wsEdit.getChangeAnnotations().values().stream().anyMatch(ca -> ca.getNeedsConfirmation())) {
+			Refactoring refactoring = new Refactoring() {
```
Could we move lines 938-973 into `runRefactorWizardOperation` passing the `change` to that method instead of the `wizard` as now?

BoykoAlex replied:
Yes, definitely!

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java:
```diff
@@ -925,25 +933,79 @@ public static void applyWorkspaceEdit(WorkspaceEdit wsEdit, String label) {
+				}
+
+			};
+			if (Display.getCurrent() == null) {
```
for consistency within this method, could we invert the condition on `Display.getCurrent()` in the new or the old code so that they use the same condition and the the if/else is in the same order?

BoykoAlex replied:
I'll create runOnUiThread(Runnable) in the UI class and use it here in both places

rubenporras reviewed (commented):
Hi Alex, 
the PR makes sense to me. I looks good to me but I do not know the Refactoring API at all.
My only comments are just some minor improvements on the code structure. If @martinlippert or @mickaelistria don't have further comments I would be happy if we merge it

mickaelistria commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java:
```diff
@@ -925,25 +933,79 @@ public static void applyWorkspaceEdit(WorkspaceEdit wsEdit, String label) {
+				}
+
+			};
+			if (Display.getCurrent() == null) {
+				UI.getDisplay().asyncExec(() -> runRefactorWizardOperation(wizard, label));
+			} else {
+				runRefactorWizardOperation(wizard, label);
 			}
```
I think it would be better to put the UI thread check directly in `runRefactorWizardOperation`.

mickaelistria reviewed (commented):
I had a quick look and it seems good overally,
I'm disappointed that there isn't a utility for that already in Platform, something like `new ChangeRefactoringWizard(change)`, nor nothing alike (as far as I search it). So in absence of such a more straightforward API, this implementation seems to use the correct bits.
I support the overall comments about factorizing more stuff into the `runRefactorWizardOperation`.

BoykoAlex commented:
@rubenporras @mickaelistria Thank you both very much for the prompt review :-)
I have added a commit to address the review comments with a few more changes:
1. Even if the workspace change is for one file that is opened in the editor in the case of `needsConfirmation` flag set to true we would still bring up the refactor preview. (I think this is rather obvious change)
2. Added `runOnUIThread(Runnable)` to the `UI` class add used it in `LSPEclipseUtils`
3. `ServerMessageHandler` use current active shell as the parent for `MessageDialog` instead of creating a new one - don't think you'd ever need to create a parent shell in Eclipse for a dialog. Creation of a new shell leaves its bounds undetermined the consequence of which is dialog box moves to bottom-right every time a new one is opened.

rubenporras commented:
thank you
